### ISSUE TITLE
ES6 class demonstration snippet remarks

### ISSUE THREAD

**Issue by mihalydanis** [getify/You-Dont-Know-JS#1207]
Please consider the following code snippet you've used as ES6 class demonstration:

```js
class P {
	foo() { console.log( "P.foo" ); }
}

class C extends P {
	foo() {
		super(); // **1**
	}
}

var c1 = new C();
c1.foo(); // "P.foo"

var D = {
	foo: function() { console.log( "D.foo" ); }
};

var E = {
	foo: C.prototype.foo
};

// Link E to D for delegation
Object.setPrototypeOf( E, D );

E.foo(); // "P.foo" **2**
```

I'd have two questions regarding this snippet; marked the exact positions:

1. Calling super() gives a SyntaxError here. _Shouldn't it be a super.foo() call?_
2. Why would you initially expect that E.foo() using super() should call to D.foo()? _I would expect it calls P.foo(), so exactly how it actually works now._

Link:
https://github.com/getify/You-Dont-Know-JS/blob/3f9efe8aefb6d2b8ff9db983802acf62f7905edb/this%20%26%20object%20prototypes/apA.md#class-gotchas

**Comment by getify** [comment]
> 1. Calling super() gives a SyntaxError here. Shouldn't it be a super.foo() call?

The spec changed subsequent to the writing of the book, this is a planned update for second edition. You're right, it must be `super.foo()`.

> 2. Why would you initially expect that E.foo() using super() should call to D.foo()?

`E`'s prototype is set to `D`, so just as `this` respects the prototype chain, I would expect `super` to do so as well. 

Setting classes aside (as they're claimed to only be syntax sugar), consider just some objects with prototype delegation:

```js
var A1 = {
   x: 1,
   foo() { return 2; }
};

var B = {
   __proto__: A1,
   bar() { console.log( this.x, super.foo() ); }
};

B.bar();   // 1 2
```

OK, that makes perfectly fine sense, right? But now...

```js
var A2 = {
   x: 3,
   foo() { return 4; }
};

// change where B delegates now
Object.setPrototypeOf( B, A2 );

B.bar();   // 3 2   <--- what!?
```

When we changed where `B` delegates to, from `A1` to `A2`, the `this.x` now finds `3` from `A2`, but the `super.foo()` is still referencing the `foo()` from `A1`.

That's quite surprising to me. I can't think of any reasoning that would expect `3 2` as the output instead of `3 4`.

> I would expect it calls P.foo(), so exactly how it actually works now.

TBH I don't understand at all why you would think that, because it doesn't fit with any precedent I'm aware of (from JS or any other language, for that matter).

In fact, based on discussions I observed at the time, the TC39 spec writers *only* did it for performance reasons (as a concession to JS engine implementors), not because it made any actual design sense to statically bind `super` instead of binding it dynamically.

I can't see any advantage to the confusion of `this` being dynamic and `super` being static. That's more complexity for people to learn, and I don't think there's a single practical benefit (other than the aforementioned performance implication from the engine perspective). I can't imagine any reason you'd want to intentionally design code to take advantage of this difference, and have the slightest inkling that future readers of that code wouldn't be surprised.

I think it's an ugly wart on JS that we ended up with it this way.

**Comment by getify** [comment]
> 1. Calling super() gives a SyntaxError here. Shouldn't it be a super.foo() call?

The spec changed subsequent to the writing of the book, this is a planned update for second edition. You're right, it must be `super.foo()`.

> 2. Why would you initially expect that E.foo() using super() should call to D.foo()?

`E`'s prototype is set to `D`, so just as `this` respects the prototype chain, I would expect `super` to do so as well. 

Setting classes aside (as they're claimed to only be syntax sugar), and even setting aside the reassignment of prototypes, consider just some simple objects with basic prototype delegation:

```js
var A = {
   x: 1,
   foo() { return 2; }
};

var B = {
   __proto__: A,
   bar() { console.log( this.x, super.foo() ); }
};

B.bar();   // 1 2

// ***************

var C = {
   x: 3,
   foo() { return 4; }
};

var D = {
   __proto__: C,
   bar() { console.log( this.x, super.foo() ); }
};

D.bar();  // 3 4
```

OK, hopefully that all makes perfectly fine sense, right? But now...

```js
B.bar.call(D);   // 3 2   <--- what!?
```

When we call `B.bar(..)` to run in the context of `D` instead of `B`, the `this.x` now finds `3` from `C` -- exactly as expected, given understanding dynamic context binding -- but the `super.foo()` is still referencing the `foo()` from `A`.

That's quite surprising to me. I can't think of any reasoning that would expect `3 2` as the output instead of `3 4`.

> I would expect it calls P.foo(), so exactly how it actually works now.

TBH I don't understand at all why you would think that, because it doesn't fit with any precedent I'm aware of (from JS or any other language, for that matter).

In fact, based on discussions I observed at the time, the TC39 spec writers *only* did it for performance reasons (as a concession to JS engine implementors), not because it made any actual design sense to statically bind `super` instead of binding it dynamically.

I can't see any advantage to the confusion of `this` being dynamic and `super` being static. That's more complexity for people to learn, and I don't think there's a single practical benefit (other than the aforementioned performance implication from the engine perspective). I can't imagine any reason you'd want to intentionally design code to take advantage of this difference, and have the slightest confidence that future readers of that code wouldn't be surprised.

I think it's an ugly wart on JS that we ended up with it this way.
